File: Top-Interview-Questions-Leetcode/Easy-Collections/linked-list/reverse-linked-list.cpp

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
private:
    
    ListNode* reverse_recursively(ListNode* head) {
        if (head->next == nullptr) return head;
        
        ListNode* reverse_sublist = reverse_recursively(head->next);
        
        head->next->next = head;
        
        head->next = nullptr;
        
        return reverse_sublist;
    }
    
    ListNode* reverse_iteratively(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* next = head;
        
        while (head) {
            next = head->next;
            head->next = prev;
            prev = head;
            head = next;
        }
        
        return prev;
    }
    
public:
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr) return head;
        
        return reverse_recursively(head);
    }
};

File: Top-Interview-Questions-Leetcode/Easy-Collections/linked-list/reverse-linked-list.py
from typing import Optional 

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def __reverse_iteratively(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        next = None
        
        while head != None:
            next = head.next
            head.next = prev
            prev = head
            head = next
        
        return prev
    
    def __reverse_recursively(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head.next == None:
            return head
        
        reverse_sublist = self.__reverse_recursively(head.next)
        
        head.next.next = head
        head.next = None
        
        return reverse_sublist
            
    
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None:
            return head
        
        return self.__reverse_recursively(head)
        